solver: ensure commit variants enter dag hash

psakievich opened the pull request:
Resolves #51167 by moving the commit variant population before specs are finalized.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

psakievich committed:
solver: ensure commit variants enter dag hash

Resolves #51167 by moving the commit variant population before specs are
finalized.

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Resolve additional failing tests

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Make private member function to API deprecation

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Whitespace error

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Fix monkeypatch calling method

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Fix more tests

Signed-off-by: psakiev <psakiev@sandia.gov>

haampie commented:
FYI, if you need a feature backported, use one or more `vX.Y.Z` labels. The [`backports` label](https://github.com/spack/spack/issues?q=label%3Abackports) is for backports PRs themselves.

alalazo commented on lib/spack/spack/test/concretization/core.py:
We have a test to unmark as xfail https://github.com/spack/spack/pull/51182

psakievich replied:
I removed the xfail decorator from that test in this PR. I'm assuming you might have missed that in the review?

alalazo replied:
Yeah, definitely. The funny thing is that I looked for it closely a couple of times

psakievich commented on lib/spack/spack/test/cmd/mirror.py:
```diff
@@ -609,7 +609,7 @@ def test_binary_provenance_url_fails_mirror_resolves_commit(
     mirror("add", "--type", "source", "test-mirror", mirror_path)
 
     spec = spack.concretize.concretize_one("git-test-commit@main")
-    assert spec.package.stage.archive_file
```
I think there still might be a bug here. Manual inspection of the archive path makes me want to confirm before merging?

psakievich replied:
Resolved concern with a new assertion

alalazo commented on lib/spack/spack/package_base.py:
```diff
@@ -1077,23 +1073,36 @@ def resolve_binary_provenance(self) -> None:
+            if pkg_instance.stage.archive_file:
+                sha = spack.util.archive.retrieve_commit_from_archive(
+                    pkg_instance.stage.archive_file, ref
+                )
```
Just for reference, for this part I started refactoring `UrlFetchStrategy` in https://github.com/spack/spack/pull/51192. 

The idea is that we don't need to do this analysis in a stage, and construct manually `pkg_cls(spec)`. What we can do is just download the archive in some temporary place and inspect it.

psakievich replied:
The goal here is to not download. I want to first look for a stage that already exists, so I want this functions stage dir to match the stage that will be computed by the package instance after the spec is finalized. I'm not sure if this is correct yet. 

But I switched to this strategy because I was not getting the mirrors and didn't want to recreate the logic for populating them when constructing the fetcher. It is embedded in the package class method.

alalazo replied:
> I want to first look for a stage that already exists, so I want this functions stage dir to match the stage that will be computed by the package instance after the spec is finalized.

This is difficult to do because :recycle: 
1. The stage directory uses the hash of the spec, 
2. The hash needs to account for the `commit=` value,
3. To compute that value we want to look into the stage.

If it's to special case `dev_path=` specs, then we could use the `dev_path=` value and otherwise download?

psakievich replied:
Yeah good point. I need to think this through. Probably the most important thing is the archive is the same stored in the local mirror. But I might just drop the stage check. I don't see how that can ever work now that you point it out.

psakievich replied:
We actually skip this check for dev_path since people developing may be changing the commit quite often without wanting the concretization to change. dev_path is its own sort of src provenance. 

The goal of checking the stage was to help reduce concretizer churn when people have to force reconcretize because it is fast local and gives another level of control beyond `--fresh`. This does add overhead to the solve and I wanted to minimize it.

Another option could be to add a pin command/flag that would record the commits in config so users can skip this check by saying preserve provenance from the last concretization but otherwise do a `--fresh` solve. 

Maybe I'm over engineering this right now though

psakievich replied:
@alalazo I removed the stage check but kept the package instance creation for the mirror population. Feel free to unresolve this thread if you want any actions from me.

psakievich committed:
Remove stage check

The stage check is no longer valid pre spec finalize. These unfinalized
specs should never have the same stage since we are adding a variant for
this method.

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Line width

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Merge branch 'develop' into hash-to-hash

psakievich committed:
Clarify docs

Move main docs to packaging and add a qualifying not in the git versions
section

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
Style

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich committed:
rst style :eye_roll:

Signed-off-by: psakiev <psakiev@sandia.gov>

becker33 commented on lib/spack/docs/packaging_guide_creation.rst:
```diff
@@ -603,6 +603,39 @@ So, if the sources are at ``http://example.com/downloads/foo-1.0.tar.gz``, Spack
 
+.. _git_version_binary_provenance:
+
+Git Version Binary Provenance
```
I think we should avoid referring to this as "binary provenance". This is probably properly considered source provenance, since it's information about what source code we're using.

psakievich replied:
To me it's both. Is there an alternative definition of binary provenance you're thinking of?

The src provenance is now embedded on the spec which then enables binary provenance i.e. this binary was built from this source. The motivation for this feature was so reuse could be provided with confidence for these binaries. I consider the stage src provenance and the installed package binary provenance. We've had src provenance for a long time since the archives are labeled by commit. 

I don't mind iterating too much. I'm just a little surprised to get the this request at this stage in the process.

becker33 replied:
"binary provenance" sounds to me like where a binary came from -- which would be null for a source build. I don't disagree at all that source provenance is all inherently part of binary provenance, I just think that by changing the name we can head off an entire realm of misunderstandings.

When we were iterating on the original feature before and it was all in my head constantly I couldn't 'misread' it in my head. With a bit more perspective coming back to it my first thought was "there's no binary involved at all" (although you're right the end artifact is itself a binary, we don't usually talk about it that way)

becker33 commented on lib/spack/docs/packaging_guide_creation.rst:
```diff
@@ -603,6 +603,39 @@ So, if the sources are at ``http://example.com/downloads/foo-1.0.tar.gz``, Spack
+
+Checksummed assets are preferred but there are a few notable exceptions such as git branches and tags i.e ``pkg@develop``.
+These versions do not naturally have source provenance because they refer to a range of commits (branches) or can be changed outside the spack packaging infrastructure (tags).
+Without source provenance we cannot have binary provenance.
```
```suggestion
```

becker33 commented on lib/spack/docs/packaging_guide_creation.rst:
```diff
@@ -603,6 +603,39 @@ So, if the sources are at ``http://example.com/downloads/foo-1.0.tar.gz``, Spack
+These versions do not naturally have source provenance because they refer to a range of commits (branches) or can be changed outside the spack packaging infrastructure (tags).
+Without source provenance we cannot have binary provenance.
+
+Spack has a reserved variant to allow users to complete source and binary provenance for these cases: ``pkg@develop commit=<SHA>``.
```
```suggestion
Spack has a reserved variant to allow users to complete source provenance for these cases: ``pkg@develop commit=<SHA>``.
```

becker33 commented on lib/spack/docs/packaging_guide_creation.rst:
```diff
@@ -603,6 +603,39 @@ So, if the sources are at ``http://example.com/downloads/foo-1.0.tar.gz``, Spack
+The ``commit`` variant must be supplied using the full 40 character commit SHA.
+Using a partial commit SHA or assigning the ``commit`` variant to a version that is not using a branch or tag reference will lead to an error during concretization.
+
+Spack will attempt to establish binary provenance by looking up commit SHA's for branch and tag based versions during concretization.
```
```suggestion
Spack will attempt to establish source provenance by looking up commit SHA's for branch and tag based versions during concretization.
```

becker33 commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -191,6 +191,13 @@ matches any version in the range ``1.0:1.5`` and the specific version ``1.7.1``.
 
+.. note::
+   Users wanting to just match specific commits for branch or tag based versions should assign the ``commit`` variant (``commit=<40 char sha>``).
+   Spack reserves this variant specifically to track binary provenance of git based versions.
```
```suggestion
   Spack reserves this variant specifically to track source provenance of git based versions.
```

becker33 commented on lib/spack/docs/spec_syntax.rst:
```diff
@@ -191,6 +191,13 @@ matches any version in the range ``1.0:1.5`` and the specific version ``1.7.1``.
+   Users wanting to just match specific commits for branch or tag based versions should assign the ``commit`` variant (``commit=<40 char sha>``).
+   Spack reserves this variant specifically to track binary provenance of git based versions.
+   Spack will attempt to compute this value for you automatically during concretization and raise a warning if it is unable to assign the commit.
+   Further details can be found in :ref:`git_version_binary_provenance`.
```
We should change the ref as well.

becker33 commented on lib/spack/spack/package_base.py:
```diff
@@ -1055,58 +1055,65 @@ def needs_commit(cls, version) -> bool:
+        pkg_instance = cls(spec)
+
+        try:
+            pkg_instance.do_fetch(mirror_only=True)
```
@psakievich have you checked whether this incidentally sets any hashes on the spec? It may be better do do `pkg_instance = cls(spec.copy())` above to ensure we don't accidentally mutate the spec through `pkg_instance` methods.

psakievich replied:
Good thought. I have not checked that. I think a copy is a fine idea. Probably the safest choice. I'll update the PR.

psakievich reviewed (commented):
Leaving a comment since my phone is making all my comments be embedded in reviews.

psakievich committed:
Respond to reviews

Signed-off-by: psakiev <psakiev@sandia.gov>

psakievich commented:
@becker33 I just removed the word `binary` from the docs and call it `git version provenance`.  I'm happy to rework the docs further if you have more concerns, but I'd like to do it in a follow on so the bug fix is live on develop.

psakievich committed:
Fix name

Signed-off-by: psakiev <psakiev@sandia.gov>
